### PR TITLE
LibWeb: Make Window.postMessage closer to the spec

### PR DESCRIPTION
**Pull request by Lubrsi**
The main issues are using Structured{Serialize,Deserailize} instead of
Structured{Serialize,Deserialize}WithTransfer and the temporary
execution context usage for StructuredDeserialize.

Allows Discord to load once again, as it uses a postMessage scheduler
to render components, including the main App component. The callback
checked the (previously) non-existent source attribute of the
MessageEvent and returned if it was not the main window.

Fixes the Twitch cookie consent banner saying "failed integrity check"
for unknown reasons, but presumably related to the source and origin
attributes.

### CODE REVIEW

File: Userland/Libraries/LibWeb/HTML/Window.cpp [hunk 1]
```diff
@@ -997,17 +998,106 @@ Optional<String> Window::prompt(Optional<String> const& message, Optional<String
...
+        // 3. Let source be the WindowProxy object corresponding to incumbentSettings's global object (a Window object).
+        auto& source = verify_cast<WindowProxy>(incumbent_settings.realm().global_environment().global_this_value());
+
+        // 4. Let deserializeRecord be StructuredDeserializeWithTransfer(serializeWithTransferResult, targetRealm).
+        // FIXME: Use StructuredDeserializeWithTransfer instead of StructuredDeserialize
+        // FIXME: Don't use a temporary execution context here.
+        auto& settings_object = Bindings::host_defined_environment_settings_object(target_realm);
+        settings_object.prepare_to_run_script();
+        auto deserialize_record_or_error = structured_deserialize(vm(), serialize_with_transfer_result, target_realm, Optional<HTML::SerializationMemory> {});
+        settings_object.clean_up_after_running_script();
```
Comment (ADKaster) [inline Userland/Libraries/LibWeb/HTML/Window.cpp:1059]: In https://github.com/SerenityOS/serenity/pull/21254 I have a commit to move the temp execution context into structured_deserialize.

Comment (trflynn89) [reply Userland/Libraries/LibWeb/HTML/Window.cpp:1059]: Not sure if this is staying here, but please use `Web::HTML::TemporaryExecutionContext` if so

File: Userland/Libraries/LibWeb/HTML/Window.cpp [hunk 2]
```diff
@@ -997,17 +999,114 @@ Optional<String> Window::prompt(Optional<String> const& message, Optional<String
...
+    Variant<String, Origin> target_origin = options.target_origin;
+
+    // 4. If targetOrigin is a single U+002F SOLIDUS character (/), then set targetOrigin to incumbentSettings's origin.
+    if (options.target_origin == "/"sv) {
+        target_origin = incumbent_settings.origin();
+    }
+    // 5. Otherwise, if targetOrigin is not a single U+002A ASTERISK character (*), then:
+    else if (options.target_origin != "*"sv) {
+        // 1. Let parsedURL be the result of running the URL parser on targetOrigin.
+        auto parsed_url = AK::URL(options.target_origin);
```
Comment (shannonbooth) [inline Userland/Libraries/LibWeb/HTML/Window.cpp:1021]: This should technically be calling `URL::parse`, AK::URL is calling basic-parse (functionally this is the same currently until we have better blob support): https://github.com/SerenityOS/serenity/blob/master/Userland/Libraries/LibWeb/URL/URL.h#L97

File: Userland/Libraries/LibWeb/HTML/Window.cpp [hunk 3]
```diff
@@ -997,17 +999,114 @@ Optional<String> Window::prompt(Optional<String> const& message, Optional<String
...
+    // 6. Let transfer be options["transfer"].
+    // FIXME: This is currently unused.
+
+    // 7. Let serializeWithTransferResult be StructuredSerializeWithTransfer(message, transfer). Rethrow any exceptions.
+    // FIXME: Use StructuredSerializeWithTransfer instead of StructuredSerialize
+    auto serialize_with_transfer_result = TRY(structured_serialize(target_realm.vm(), message));
+
+    // 8. Queue a global task on the posted message task source given targetWindow to run the following steps:
+    queue_global_task(Task::Source::PostedMessage, *this, [this, serialize_with_transfer_result = move(serialize_with_transfer_result), target_origin = move(target_origin), &incumbent_settings, &target_realm]() {
+        // 1. If the targetOrigin argument is not a single literal U+002A ASTERISK character (*) and targetWindow's
```
Comment (shannonbooth) [inline Userland/Libraries/LibWeb/HTML/Window.cpp:1040]: I might be missing something, but this implementation doesn't seem to match the spec

**Review by shannonbooth — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibWeb/HTML/Window.cpp (hunk 3 above)
Comment (Lubrsi) [reply Userland/Libraries/LibWeb/HTML/Window.cpp:1040]: AFAICT it's only a `String` if that `String` is '*' due to steps 4 and 5 always setting `targetOrigin` to an `Origin` if `targetOrigin` is not `*', added a NOTE clarification.

File: Userland/Libraries/LibWeb/HTML/Window.cpp (hunk 2 above)
Comment (Lubrsi) [reply Userland/Libraries/LibWeb/HTML/Window.cpp]: Ah, didn't know about that :^), this prompted to add a test of postMessage from a blob URL iframe

File: Userland/Libraries/LibWeb/HTML/Window.cpp (hunk 3 above)
Comment (shannonbooth) [reply Userland/Libraries/LibWeb/HTML/Window.cpp:1040]: Right! Yup, that makes sense

**Pull request merged**